File: game/post_processing.py
import pygame

from config import Config
from tween import Tween


class PostProcessing:
    def __init__(self):
        self.wavyness = 0
        self.bitcrush = 256
        self.hue_shift = 0
        self.scale = 1
        self.line_glitch = 0
        self.chromatic_aberration = 0
        pygame.joystick.init()
        self.joysticks = [pygame.joystick.Joystick(x) for x in range(pygame.joystick.get_count())]

    def handle_events(self):
        if len(self.joysticks) == 0:
            return

        if self.joysticks[0].get_numaxes() != 6:
            return

        self.scale = Tween.map(self.joysticks[0].get_axis(0), -1, 1, 1, 4)
        Config.FOREST_BG_SPEED_MULTIPLIER = Tween.map(self.joysticks[0].get_axis(1), -1, 1, 1, 5)
        self.wavyness = Tween.map(self.joysticks[0].get_axis(2), -1, 1, 0, 0.1)
        self.hue_shift = Tween.map(self.joysticks[0].get_axis(3), -1, 1, 0, 1)
        self.bitcrush = Tween.map_bitcrush(self.joysticks[0].get_axis(4), -1, 1, 2, 256)
        self.line_glitch = Tween.map(self.joysticks[0].get_axis(4), -1, 1, 0, 0.5)
        self.chromatic_aberration = Tween.map(self.joysticks[0].get_axis(5), -1, 1, 0, 1)

    def apply(self, program, any_playing):
        if any_playing:
            program['wavyness'] = self.wavyness
            program['bitcrush'] = self.bitcrush
            program['hue_shift'] = self.hue_shift
            program['scale'] = self.scale
            program['chromatic_aberration'] = self.chromatic_aberration
            program['line_glitch'] = self.line_glitch
        else:
            program['wavyness'] = 0.01
            program['bitcrush'] = 256
            program['hue_shift'] = 0
            program['scale'] = 1
            program['chromatic_aberration'] = 0
            program['line_glitch'] = 0

File: game-web/src/postProcessing.ts
/**
 * Port of game/post_processing.py.
 * Manages WebGL2 post-processing shader uniforms.
 * Joystick handled via Gamepad API (replaces pygame joystick).
 */

import { Config } from './config';
import { Tween } from './tween';

export class PostProcessing {
  wavyness = 0;
  bitcrush = 256;
  hueShift = 0;
  scale = 1;
  lineGlitch = 0;
  chromaticAberration = 0;

  /** When true, slider values are used instead of the gamepad. */
  slidersActive = false;

  handleEvents(): void {
    // Sliders take priority — skip gamepad when active
    if (this.slidersActive) return;

    const gamepads = navigator.getGamepads();
    const joy = gamepads[0];
    if (!joy) return;
    if (joy.axes.length < 6) return;

    this.scale                  = Tween.map(joy.axes[0], -1, 1, 1, 4);
    Config.FOREST_BG_SPEED_MULTIPLIER = Tween.map(joy.axes[1], -1, 1, 1, 5);
    this.wavyness               = Tween.map(joy.axes[2], -1, 1, 0, 0.1);
    this.hueShift               = Tween.map(joy.axes[3], -1, 1, 0, 1);
    this.bitcrush               = Tween.mapBitcrush(joy.axes[4], -1, 1, 2, 256);
    this.lineGlitch             = Tween.map(joy.axes[4], -1, 1, 0, 0.5);
    this.chromaticAberration    = Tween.map(joy.axes[5], -1, 1, 0, 1);
  }

  apply(gl: WebGL2RenderingContext, program: WebGLProgram, time: number, anyPlaying: boolean): void {
    const u = (name: string) => gl.getUniformLocation(program, name);

    gl.uniform1f(u('time'), time);

    if (anyPlaying) {
      gl.uniform1f(u('wavyness'),            this.wavyness);
      gl.uniform1f(u('bitcrush'),            this.bitcrush);
      gl.uniform1f(u('hue_shift'),           this.hueShift);
      gl.uniform1f(u('scale'),               this.scale);
      gl.uniform1f(u('chromatic_aberration'), this.chromaticAberration);
      gl.uniform1f(u('line_glitch'),         this.lineGlitch);
    } else {
      gl.uniform1f(u('wavyness'),            0.01);
      gl.uniform1f(u('bitcrush'),            256);
      gl.uniform1f(u('hue_shift'),           0);
      gl.uniform1f(u('scale'),               1);
      gl.uniform1f(u('chromatic_aberration'), 0);
      gl.uniform1f(u('line_glitch'),         0);
    }
  }
}
